### PR TITLE
Citations Updates 2 - Remix Edition

### LINKED ISSUE #45004: Silicons can't use security record consoles
## Testmerges:
#43148 
#44945 

## Reproduction:
1. Be an AI
2. Log into a security records console
3. Try to add a crime (citation/minor/major)
4. Get the "_You are too far away_" message

Unsure what exactly is causing this. I made some changes to the security record consoles in #44853 but I shouldn't have touched anything related to silicon use.

### LINKED ISSUE #45259: Security warrant computer missing multiple sanity checks for login
Master.

Doesn't check for null ID's
doesn't check for silicons
Doesn't check for adminghosts

```
[2019-07-18 07:44:42.379] runtime error: Cannot read null.registered_name
 - proc name: Topic (/obj/machinery/computer/warrant/Topic)
 -   source file: warrant.dm,118
 -   usr: M.R. B.U.B.B.L.E.S (/mob/living/silicon/robot)
 -   src: the security warrant console (/obj/machinery/computer/warrant)
 -   usr.loc: the floor (109,160,2) (/turf/open/floor/plasteel)
 -   src.loc: the floor (111,158,2) (/turf/open/floor/plasteel)
 -   call stack:
 - the security warrant console (/obj/machinery/computer/warrant): Topic("src=\[0x2007fb4];choice=Login", /list (/list))
 - Kidrik (/client): Topic("src=\[0x2007fb4];choice=Login", /list (/list), the security warrant console (/obj/machinery/computer/warrant))
 - Kidrik (/client): Topic("src=\[0x2007fb4];choice=Login", /list (/list), the security warrant console (/obj/machinery/computer/warrant))
 - 
```

https://github.com/tgstation/tgstation/blob/5e4dafb30e653ea3491f06be0a12037390eab9a4/code/game/machinery/computer/warrant.dm#L116-L123

Example for how security record consoles work
https://github.com/tgstation/tgstation/blob/5e4dafb30e653ea3491f06be0a12037390eab9a4/code/game/machinery/computer/security.dm#L348-L368

### PR DESCRIPTION
**Pull request by nfreader**
## About The Pull Request
I just love touching SQL! This is a second pass on citations, adding support for adding them via HUDs, alerting security when they get paid, and most excitingly, tracking citations in the database. This also fixes some bugs I ran into earlier, as well as lays some groundwork for refactoring sec/med record consoles.

This PR should have substantially LESS commit thrashing.

## Why It's Good For The Game
Tracking citations on the DB was the plan all along and you've fallen into my nefarious trap! 

Please direct all feedback/ideas to this thread: https://tgstation13.org/phpBB/viewtopic.php?f=9&t=20083 and reserve comments here for discussing code changes.

## Changelog
:cl: Atlanta Ned
add: Security citations are now tracked in the database!
add: Security can now cite people from their HUD
add: Paid citations 
fix: Fixes #45004 and some other issues
/:cl:

Closes #45259

### CODE REVIEW

File: code/controllers/subsystem/blackbox.dm [hunk 1]
```diff
@@ -335,3 +335,55 @@ Versioning
...
+	if(SSdbcore.db_major < 5 && SSdbcore.db_minor < 4)
+		return
+
+	var/dataId = sanitizeSQL(id)
+
+	var/datum/DBQuery/query_delete_citation = SSdbcore.NewQuery("UPDATE [format_table_name("citations")] SET DELETED = 1 WHERE round_id = [GLOB.round_id] AND dataId = [dataId]")
+	query_delete_citation.Execute()
+	qdel(query_delete_citation)
+
+/datum/controller/subsystem/blackbox/proc/PayCitation(id, amount)
```
Comment (nfreader) [inline code/controllers/subsystem/blackbox.dm]: This is only fired when the citation is paid in full, not every time money is paid towards the citation.

File: code/game/machinery/computer/warrant.dm [hunk 2]
```diff
@@ -115,11 +121,9 @@
 	switch(href_list["choice"])
 		if("Login")
 			var/obj/item/card/id/scan = M.get_idcard(TRUE)
-			authenticated = scan.registered_name
-			if(authenticated)
-				for(var/datum/data/record/R in GLOB.data_core.security)
-					if(R.fields["name"] == authenticated)
-						current = R
+			if(scan && istype(scan))
+				authenticated = scan.registered_name
```
Comment (ShizCalev) [inline code/game/machinery/computer/warrant.dm]: #45259

File: code/game/machinery/computer/warrant.dm [hunk 3]
```diff
@@ -115,11 +121,9 @@
 	switch(href_list["choice"])
 		if("Login")
 			var/obj/item/card/id/scan = M.get_idcard(TRUE)
-			authenticated = scan.registered_name
-			if(authenticated)
-				for(var/datum/data/record/R in GLOB.data_core.security)
-					if(R.fields["name"] == authenticated)
-						current = R
+			if(scan && istype(scan))
```
Comment (ShizCalev) [inline code/game/machinery/computer/warrant.dm]: can just `istype(scan)`

File: code/game/machinery/computer/warrant.dm [hunk 4]
```diff
@@ -8,6 +8,12 @@
 	var/authenticated = null
 	var/screen = null
 	var/datum/data/record/current = null
+	var/obj/item/radio/Radio
+
+/obj/machinery/computer/warrant/Initialize()
+	. = ..()
+	Radio = new/obj/item/radio(src)
+	Radio.listening = 0
```
Comment (ShizCalev) [inline code/game/machinery/computer/warrant.dm]: FALSE

File: code/modules/mob/living/carbon/human/human.dm [hunk 5]
```diff
@@ -416,6 +416,42 @@
...
+											var/crime = GLOB.data_core.createCrimeEntry(t1, "", allowed_access, station_time_timestamp(), fine)
+											for (var/obj/item/pda/P in GLOB.PDAs)
+												if(P.owner == R.fields["name"])
+													var/message = "You have been fined [fine] credits for '[t1]'. Fines may be paid at security."
+													var/datum/signal/subspace/messaging/pda/signal = new(src, list(
+														"name" = "Security Citation",
+														"job" = "Citation Server",
+														"message" = message,
+														"targets" = list("[P.owner] ([P.ownjob])"),
+														"automated" = 1
```
Comment (ShizCalev) [inline code/modules/mob/living/carbon/human/human.dm]: TRUE

File: code/game/machinery/computer/security.dm [hunk 6]
```diff
@@ -382,7 +388,42 @@ What a mess.*/
...
+					var/crime = GLOB.data_core.createCrimeEntry(t1, "", authenticated, station_time_timestamp(), fine)
+					for (var/obj/item/pda/P in GLOB.PDAs)
+						if(P.owner == active1.fields["name"])
+							var/message = "You have been fined [fine] credits for '[t1]'. Fines may be paid at security."
+							var/datum/signal/subspace/messaging/pda/signal = new(src, list(
+								"name" = "Security Citation",
+								"job" = "Citation Server",
+								"message" = message,
+								"targets" = list("[P.owner] ([P.ownjob])"),
+								"automated" = 1
```
Comment (ShizCalev) [inline code/game/machinery/computer/security.dm]: TRUE

File: code/game/machinery/computer/warrant.dm (hunk 3 above)
Comment (ShizCalev) [reply code/game/machinery/computer/warrant.dm]: also silicons can't issue citations :(

**Pull request closed without merging**